File: LeetCode/C++/trees-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {    
public:
    bool hasPathSumRecursive(TreeNode* root, int targetSum, int sum) {
        int sum_local;
        sum += root->val;
        
        if (!root->left && !root->right) {  // leaf node case
            return (sum == targetSum ? true : false);
        }
        bool hasPathSum_l = false, hasPathSum_r = false;
        sum_local = sum;

        hasPathSum_l = root->left ? hasPathSumRecursive(root->left, targetSum, sum) : false;
        sum = sum_local;    // resetting static sum value to the local sum of node

        hasPathSum_r = root->right ? hasPathSumRecursive(root->right, targetSum, sum) : false;
        sum = sum_local;

        return hasPathSum_l || hasPathSum_r;         
    }
        
    bool hasPathSum(TreeNode* root, int targetSum) {
        return (root ? hasPathSumRecursive(root, targetSum, 0) : false);
    }
};

File: LeetCode/Python/trees-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: TreeNode, targetSum: int) -> bool:
        return self.hasPathSumRecursive(root, targetSum, 0) if root else False
        
        
    def hasPathSumRecursive(self, root: TreeNode, targetSum: int, pathSum: int) -> bool:
        pathSum += root.val
        
        if not root.left and not root.right:  # leaf node case
            return True if pathSum == targetSum else False
        
        hasPathSum_l = hasPathSum_r = False
        sum_local = pathSum
        
        hasPathSum_l = self.hasPathSumRecursive(root.left, targetSum, pathSum) if root.left \
                        else False
        pathSum = sum_local  # resetting total path sum value to the local sum of node
        
        hasPathSum_r = self.hasPathSumRecursive(root.right, targetSum, pathSum) if root.right \
                        else False
        pathSum = sum_local
        
        return hasPathSum_l or hasPathSum_r
